[Jian Yi] iP

jian7490 opened the pull request:
# ChatBaby

> “The key is not to prioritize what's on your schedule, but to schedule your priorities." – Stephen Covey

ChatBaby is a **Java-based** task management chatbot designed to help you stay on top of your tasks. It's:

- Text-based
- Easy to use
- Fast 🚀

## How to get started:

1. Download the latest release [here](https://github.com/jian7490/ip/releases/tag/v1.0).
2. Double-click the jar file to launch.
3. Start adding your tasks!

### Features:

- [x] Task management (ToDos, Deadlines, Events)
- [ ] Reminders (coming soon)
- [ ] Customizable themes (coming soon)

This is the main method:

```java
public class ChatBaby {
    public static void main(String[] args) {
        Locale.setDefault(Locale.ENGLISH);
        new ChatBaby().run();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

jian7490 committed:
level 0 implemented

jian7490 committed:
unknown handle of files

jian7490 committed:
level 1 implemented

jian7490 committed:
level 2 implemented

jian7490 committed:
level 3 implemented with the optional unmark function

jian7490 committed:
level 4 implemented; all responsibilities of the chatbox become functions

jian7490 committed:
level 5 implemented; I have to re-clone it to another source path as the current path involves Chinese characters

jian7490 committed:
level 6 implemented

jian7490 committed:
semi-automation of testing is achieved

jian7490 committed:
Enum is used

jian7490 committed:
Implement level 7

jian7490 committed:
Revert "Implement level 7"

This reverts commit 3726be39154307782c4f6f03f388feca09e30472.

jian7490 committed:
Revert "Revert "Implement level 7""

This reverts commit 00a17acd73ee27e3bffab85f2bcea0e469e2546b.

jian7490 committed:
Change deadline as dateTime objects

jian7490 committed:
Merge branch 'branch-level-7' into branch-level-8

jian7490 committed:
Include LocalDateTime objects into Deadline and Event

jian7490 committed:
Implement level 8: tasks that end on a given date can be printed

jian7490 committed:
Merge branch 'branch-level-8'

jian7490 committed:
Make the whole structure more OOP-ed.

jian7490 committed:
Merge branch 'A-MoreOOP'

jian7490 committed:
Package everything

jian7490 committed:
Merge tag 'A-Packages'

jian7490 committed:
Include Gradle to build the project

jian7490 committed:
Change the file paths

jian7490 committed:
Use JUnit to test Ui and TaskList

jian7490 committed:
Create Jar file

jian7490 committed:
Include Javadoc

jian7490 committed:
Merge tag 'A-Javadoc'

jian7490 committed:
Follow the coding standard

jian7490 committed:
Follow the coding standard

jian7490 committed:
Implement level 9: test FindCommand

jian7490 committed:
Merge branch 'branch-level-9'

# Conflicts:
#	src/main/java/chatbaby/Command.java
#	src/main/java/chatbaby/DeleteCommand.java
#	src/main/java/chatbaby/ExitCommand.java
#	src/main/java/chatbaby/ListCommand.java
#	src/main/java/chatbaby/MarkCommand.java
#	src/main/java/chatbaby/Parser.java
#	src/main/java/chatbaby/ToDo.java
#	src/main/java/chatbaby/UnmarkCommand.java

jian7490 committed:
Implement level 9: test FindCommand

matchastack commented on src/main/java/chatbaby/AddCommand.java:
```diff
@@ -0,0 +1,98 @@
+
+            default:
+                throw new ChatBabyException("Oh no!!! I'm sorry, but I don't understand that command.");
+        }
```
`switch-case` statements that is not arrow function should be indented at the same level according to the coding standard for this class.

matchastack commented on src/main/java/chatbaby/AddCommand.java:
```diff
@@ -0,0 +1,98 @@
+            case TODO -> prefixLength = 5;
+            case DEADLINE -> prefixLength = 9;
+            case EVENT -> prefixLength = 6;
+        }
```
nice to see that you have followed the consensus on forum for `switch-case` statements involving arrow `->` functions!

matchastack commented on src/main/java/chatbaby/AddCommand.java:
```diff
@@ -0,0 +1,98 @@
+        tasks.addTask(newTask);
+        System.out.println("Got it. I've added this task:\n"
+                + newTask.toString() + "\n"
+                + "Now you have " + tasks.size() + " tasks in the list.");
```
appropriate line break which makes it very easy to visualise the output of the program!

matchastack commented on src/main/java/chatbaby/Parser.java:
```diff
@@ -0,0 +1,45 @@
+            case "find":
+                return new FindCommand(curCommand);
+            default:
+                throw new ChatBabyException("I'm sorry, but I don't know what that means.");
```
... I noticed the same issue at other places too.

matchastack commented on src/main/java/chatbaby/Task.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @return A string representation of the task in file format.
+     */
+    public abstract String toFileText();
```
nit: I think this is fine an perfectly understandable. I just think `toTextFile()` might be more smooth when reading.

matchastack commented on src/main/java/chatbaby/TaskList.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    public boolean listTasksOn(LocalDate date) {
+        boolean hasTask = false;
+        System.out.println("Tasks that end on " + date.format(DateTimeFormatter.ofPattern("MMM dd yyyy")) + ":");
```
nit: This is under 120 hard-limit for character count but I think having a line break will make it easier to read without needing a horizontal scroll

matchastack commented on src/test/java/chatbaby/UiTest.java:
```diff
@@ -0,0 +1,27 @@
+package chatbaby;
+
+import org.junit.jupiter.api.Test;
+import java.io.ByteArrayOutputStream;
```
there should be an empty line between the import statements

matchastack commented on src/test/java/chatbaby/UiTest.java:
```diff
@@ -0,0 +1,27 @@
+import java.io.PrintStream;
+import java.text.Normalizer;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
static import should be placed before import

matchastack commented on src/test/java/chatbaby/TaskListTest.java:
```diff
@@ -0,0 +1,17 @@
+package chatbaby;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
... I noticed the same issue at some other places too

matchastack reviewed (commented):
Overall, the coding standard has been followed except for a few areas. 
Great job on the variable namings! I think the code is really well written is easy to follow. 
Also, big kudos on including JavaDoc for all classes and most of the public methods.

jian7490 committed:
Use Checkstyle from AddressBook and correct all code styling issues

jian7490 committed:
Merge branch 'A-CheckStyle'

jian7490 committed:
Add checkstyle files

jian7490 committed:
Implement level 10: make GUI available for ChatBaby

jian7490 committed:
Merge branch 'branch-level-10-GUI'

YINZE66 commented on src/main/java/chatbaby/TaskList.java:
```diff
@@ -0,0 +1,103 @@
+     * @param index The index of the task to retrieve.
+     * @return The task at the specified index.
+     */
+    public Task taskAt(int index) {
+        return tasks.get(index);
+    }
```
Maybe change the method name to a verb? 👀

YINZE66 commented on src/test/java/chatbaby/UiTest.java:
```diff
@@ -0,0 +1,27 @@
+public class UiTest {
+
+    @Test
+    public void greetTest() {
```
Maybe should change method name to testGreet? ❓

YINZE66 commented on src/main/java/chatbaby/AddCommand.java:
```diff
@@ -0,0 +1,101 @@
+                        + type.name().toLowerCase() + " cannot be empty.");
+            }
+            break;
+        case EVENT:
+            String[] eventParts = commandBody.split("/from ");
+            if (eventParts.length == 2) {
+                String name = eventParts[0].trim();
+                String[] eventTimes = eventParts[1].split("/to ");
+                if (eventTimes.length == 2) {
+                    try {
+                        String from = eventTimes[0].trim();
+                        String to = eventTimes[1].trim();
+                        newTask = new Event(name, from, to);
+                    } catch (DateTimeParseException e) {
+                        throw new ChatBabyException("Invalid date format. Please use yyyy-MM-dd HH:mm.");
+                    }
+                } else {
+                    throw new ChatBabyException("Oh no!!! The description of a "
+                            + type.name().toLowerCase() + " cannot be empty.");
+                }
+            } else {
+                throw new ChatBabyException("Oh no!!! The description of a "
+                        + type.name().toLowerCase() + " cannot be empty.");
+            }
+            break;
```
The indentation used here is perfect! Good job! 👍

YINZE66 commented on src/main/java/chatbaby/ChatBaby.java:
```diff
@@ -0,0 +1,115 @@
+    private TaskList tasks;
+    private Ui ui;
+
+    /**
+     * Constructs a ChatBaby instance and initializes the storage, UI, and task list.
+     * Loads tasks from the storage file, or creates a new empty task list if the file cannot be loaded.
+     */
```
The comment here might be too long? Can try split into more lines 🚀

YINZE66 reviewed (commented):
Overall, your code follows code quality well! Just a few nits I found about method naming. Nice job!! 💯

jian7490 committed:
Correct some styling issues
Change gradle settings to make sure all output are in English

jian7490 committed:
Update README.md

jian7490 committed:
Update README.md

jian7490 committed:
Update README.md

jian7490 committed:
Use assertions for Error Handling

Several classes now have assertion statements for error handling.

Such assertions help verifying certain assumptions made in code.

jian7490 committed:
Improve code quality

Names of variables and methods are changed. Magic literals are converted to constants. Code logic is improved for happy paths.

jian7490 committed:
Merge pull request #1 from jian7490/branch-A-Assertions

Use assertions for Error Handling

jian7490 committed:
Merge branch 'master' into branch-A-CodeQuality

jian7490 committed:
Merge pull request #2 from jian7490/branch-A-CodeQuality

Improve code quality

jian7490 committed:
Include sort function into the task management.

TaskList class automatically sorts the tasks based on their deadlines or ending times if having any before listing the tasks.

jian7490 committed:
Delete META-INF directory

jian7490 committed:
Change comment phrasing for Main method in Chatbaby

This helps adhere with the coding standards.

jian7490 committed:
Merge remote-tracking branch 'origin/master'

jian7490 committed:
Change variable modifiers

There are fields in certain methods for which
modifiers can be improved with private or final.

Standardizing the modifiers improves code quality.

jian7490 committed:
Refactor execute() method in AddCommand class

The execute method is too long and carries adding task
functions for all three types. It can be refactored
with helper methods for creation of each type of tasks.

Refactoring a too long method further abstracts the
implementation.

jian7490 committed:
Update userImage and babyImage field in MainWindow

The two images are declared as field read from a certain
path, but in case the path does not exist or image files
are not working, the loadImage() method complements with a
transparent image.

This is applied so that it is promised the software works
whether given the presence of the images.

jian7490 committed:
Optimize import order to align with checkStyle

jian7490 committed:
Abstract findTasks() method from execute() in FindCommand class

The execute() method in FindCommand class carries the function
of finding tasks in too much detail.

This change helps maintain the single level of abstraction
principle.

jian7490 committed:
Resolve "unsupported JavaFX configuration" warning

jian7490 committed:
Improve the GUI

jian7490 committed:
Organize import for Storage class to avoid wildcard

jian7490 committed:
Set title for the main window using JavaFX

jian7490 committed:
Include Ui.png in docs

jian7490 committed:
Resolve Java version and JavaFX version to clear warnings

jian7490 committed:
AddCommand: catch empty description exception

The execute() method in AddCommand needs to deal with the cases
when the user is not giving enough information for creating a task.

Handling possible exceptions caused by user makes sure the
program work.

jian7490 committed:
Handle errors in commands

The execute() method in different subclasses of Command needs to
deal with the cases when the user is giving the right command.

Handling possible exceptions caused by user makes sure the
program work.

jian7490 committed:
Split long lines into short ones in fxml files

jian7490 committed:
Fix the bug of detecting valid command in AddCommand

jian7490 committed:
Make "bye" command terminate the program

jian7490 committed:
Fix the bug of index out of bound for task index

The execute() methods in Mark, Unmark, and Delete Commands need to
read task index and handle possible errors of index out of bounds
that result from invalid task index the user uses in the command.

This change fixes the bug of not recognizing invalid index.

jian7490 committed:
Fix the bug of including spaces as command

The command parser automatically trims the leading and
trailing space for a command to keep the meaningful part.

This change enables the parser to understand a more variety
of command formats.

jian7490 committed:
Improve the readability of task descriptions

jian7490 committed:
Update README.md

jian7490 committed:
Merge branch 'master' of https://github.com/jian7490/ip
